File: c/mipmap.c
// Demonstrates using mipmaps for texture maps.
// To overtly show the effect of mipmaps, each mipmap reduction level has a solidly colored, contrasting texture image.
// Thus, the quadrilateral is drawn with several different colors.

#include <GL/glut.h>
#include <stdlib.h>

GLubyte mipmapImage32[32][32][4];
GLubyte mipmapImage16[16][16][4];
GLubyte mipmapImage8[8][8][4];
GLubyte mipmapImage4[4][4][4];
GLubyte mipmapImage2[2][2][4];
GLubyte mipmapImage1[1][1][4];

static GLuint texName;

void makeImages(void)
{
	int i, j;

	for(i = 0; i < 32; i++)
	{
		for(j = 0; j < 32; j++)
		{
			mipmapImage32[i][j][0] = 255;
			mipmapImage32[i][j][1] = 255;
			mipmapImage32[i][j][2] = 0;
			mipmapImage32[i][j][3] = 255;
		}
	}

	for(i = 0; i < 16; i++)
	{
		for(j = 0; j < 16; j++)
		{
			mipmapImage16[i][j][0] = 255;
			mipmapImage16[i][j][1] = 0;
			mipmapImage16[i][j][2] = 255;
			mipmapImage16[i][j][3] = 255;
		}
	}

	for(i = 0; i < 8; i++)
	{
		for(j = 0; j < 8; j++)
		{
			mipmapImage8[i][j][0] = 255;
			mipmapImage8[i][j][1] = 0;
			mipmapImage8[i][j][2] = 0;
			mipmapImage8[i][j][3] = 255;
		}
	}

	for(i = 0; i < 4; i++)
	{
		for(j = 0; j < 4; j++)
		{
			mipmapImage4[i][j][0] = 0;
			mipmapImage4[i][j][1] = 255;
			mipmapImage4[i][j][2] = 0;
			mipmapImage4[i][j][3] = 255;
		}
	}

	for(i = 0; i < 2; i++)
	{
		for(j = 0; j < 2; j++)
		{
			mipmapImage2[i][j][0] = 0;
			mipmapImage2[i][j][1] = 0;
			mipmapImage2[i][j][2] = 255;
			mipmapImage2[i][j][3] = 255;
		}
	}

	mipmapImage1[0][0][0] = 255;
	mipmapImage1[0][0][1] = 255;
	mipmapImage1[0][0][2] = 255;
	mipmapImage1[0][0][3] = 255;
}

void init(void)
{
	glEnable(GL_DEPTH_TEST);
	glShadeModel(GL_FLAT);
	glTranslatef(0.0, 0.0, -3.6);
	makeImages();
	glPixelStorei(GL_UNPACK_ALIGNMENT, 1);
	glGenTextures(1, &texName);
	glBindTexture(GL_TEXTURE_2D, texName);
	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_REPEAT);
	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT);
	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST_MIPMAP_NEAREST);
	glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, 32, 32, 0, GL_RGBA, GL_UNSIGNED_BYTE, mipmapImage32);
	glTexImage2D(GL_TEXTURE_2D, 1, GL_RGBA, 16, 16, 0, GL_RGBA, GL_UNSIGNED_BYTE, mipmapImage16);
	glTexImage2D(GL_TEXTURE_2D, 2, GL_RGBA, 8, 8, 0, GL_RGBA, GL_UNSIGNED_BYTE, mipmapImage8);
	glTexImage2D(GL_TEXTURE_2D, 3, GL_RGBA, 4, 4, 0, GL_RGBA, GL_UNSIGNED_BYTE, mipmapImage4);
	glTexImage2D(GL_TEXTURE_2D, 4, GL_RGBA, 2, 2, 0, GL_RGBA, GL_UNSIGNED_BYTE, mipmapImage2);
	glTexImage2D(GL_TEXTURE_2D, 5, GL_RGBA, 1, 1, 0, GL_RGBA, GL_UNSIGNED_BYTE, mipmapImage1);
	glTexEnvf(GL_TEXTURE_ENV, GL_TEXTURE_ENV_MODE, GL_DECAL);
	glEnable(GL_TEXTURE_2D);
}

void display(void)
{
	glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
	glBindTexture(GL_TEXTURE_2D, texName);
	glBegin(GL_QUADS);
	glTexCoord2f(0.0, 0.0);
	glVertex3f(-2.0, -1.0, 0.0);
	glTexCoord2f(0.0, 8.0);
	glVertex3f(-2.0, 1.0, 0.0);
	glTexCoord2f(8.0, 8.0);
	glVertex3f(2000.0, 1.0, -6000.0);
	glTexCoord2f(8.0, 0.0);
	glVertex3f(2000.0, -1.0, -6000.0);
	glEnd();
	glFlush();
}

void reshape(int w, int h)
{
	glViewport(0, 0, (GLsizei) w, (GLsizei) h);
	glMatrixMode(GL_PROJECTION);
	glLoadIdentity();
	gluPerspective(60.0, (GLfloat)w/(GLfloat)h, 1.0, 30000.0);
	glMatrixMode(GL_MODELVIEW);
	glLoadIdentity();
}

void keyboard(unsigned char key, int x, int y)
{
	switch(key)
	{
		case 27:
			exit(0);
			break;

		default:
			break;
	}
}

int main(int argc, char** argv)
{
	glutInit(&argc, argv);
	glutInitDisplayMode(GLUT_SINGLE | GLUT_RGB | GLUT_DEPTH);
	glutInitWindowSize(500, 500);
	glutInitWindowPosition(50, 50);
	glutCreateWindow(argv[0]);
	init();
	glutDisplayFunc(display);
	glutReshapeFunc(reshape);
	glutKeyboardFunc(keyboard);
	glutMainLoop();
	return 0;
}

File: python/Mipmap.py
#!/usr/bin/python

from Window import *

class Mipmap(Window):
	''' Demonstrates using mipmaps for texture maps.'''

	def __init__(self):
		super(Mipmap, self).__init__("mipmap.c", "Mipmap", 500, 500)
		glutInitDisplayMode(GLUT_SINGLE | GLUT_RGB | GLUT_DEPTH)
		glEnable(GL_DEPTH_TEST)
		glShadeModel(GL_FLAT)
		glTranslatef(0, 0, -3.6)
		self.makeImages()
		glPixelStorei(GL_UNPACK_ALIGNMENT, 1)
		glGenTextures(1, self.texName)
		self.texName = glBindTexture(GL_TEXTURE_2D)
		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_REPEAT)
		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT)
		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST)
		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST_MIPMAP_NEAREST)
		glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, 32, 32, 0, GL_RGBA, GL_UNSIGNED_BYTE, self.mipmap32)
		glTexImage2D(GL_TEXTURE_2D, 1, GL_RGBA, 16, 16, 0, GL_RGBA, GL_UNSIGNED_BYTE, self.mipmap16)
		glTexImage2D(GL_TEXTURE_2D, 2, GL_RGBA,  8,  8, 0, GL_RGBA, GL_UNSIGNED_BYTE, self.mipmap08)
		glTexImage2D(GL_TEXTURE_2D, 3, GL_RGBA,  4,  4, 0, GL_RGBA, GL_UNSIGNED_BYTE, self.mipmap04)
		glTexImage2D(GL_TEXTURE_2D, 4, GL_RGBA,  2,  2, 0, GL_RGBA, GL_UNSIGNED_BYTE, self.mipmap02)
		glTexImage2D(GL_TEXTURE_2D, 5, GL_RGBA,  1,  1, 0, GL_RGBA, GL_UNSIGNED_BYTE, self.mipmap01)
		glTexEnvf(GL_TEXTURE_ENV, GL_TEXTURE_ENV_MODE, GL_DECAL)
		glEnable(GL_TEXTURE_2D)

	def display(self):
		glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT)
		glBindTexture(GL_TEXTURE_2D, texName)
		glBegin(GL_QUADS)
		glTexCoord2f(0, 0)
		glVertex3f(-2, -1, 0)
		glTexCoord2f(0, 8)
		glVertex3f(-2, 1, 0)
		glTexCoord2f(8, 8)
		glVertex3f(2000, 1, -6000)
		glTexCoord2f(8, 0)
		glVertex3f(2000, -1, -6000)
		glEnd()
		glFlush()

	def makeImages(self):
		''' To overtly show the effect of mipmaps, each mipmap reduction level has a solidly colored, contrasting texture image.'''
		''' Thus, the quadrilateral is drawn with several different colors.'''
		self.mipmap32 = bytes[32][32]
		self.mipmap16 = bytes[16][16]
		self.mipmap08 = bytes[ 8][ 8]
		self.mipmap04 = bytes[ 4][ 4]
		self.mipmap02 = bytes[ 2][ 2]
		self.mipmap01 = bytes[ 1][ 1]
		for i in range(32):
			for j in range(32):
				self.mipmap32[i][j] = (255, 255,   0, 255)
		for i in range(16):
			for j in range(16):
				self.mipmap16[i][j] = (255,   0, 255, 255)
		for i in range(8):
			for i in range(8):
				self.mipmap08[i][j] = (255,   0,   0, 255)
		for i in range(4):
			for i in range(4):
				self.mipmap04[i][j] = (  0, 255,   0, 255)
		for i in range(2):
			for i in range(2):
				self.mipmap02[i][j] =  (  0,  0, 255, 255)
		for i in range(1):
			for i in range(1):
				self.mipmap01[0][0] = (255, 255, 255, 255)

if __name__ == '__main__':
	Mipmap().run()
